include/intercept.h

```diff
@@ -0,0 +1,93 @@
+#ifndef _INTERCEPT_H
+#define _INTERCEPT_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdarg.h>
+
+// vDSO wrapper needed includes
+#include <time.h>
+
+#include "cheriintrin.h"
+
+#include "mem_mng.h"
+
+// Forward declarations
+struct Compartment;
+extern struct Compartment* loaded_comp;
+int64_t exec_comp(struct Compartment*, char*, char**);
+struct Compartment* manager_get_compartment_by_id(size_t);
+
+extern void* __capability manager_ddc;
+
+// Number of capabilities required to perform a transition
+#define COMP_RETURN_CAPS_COUNT 2
+
+// Capabilities required to transition back into the manager once compartment
+// execution has finished
+extern void* __capability comp_return_caps[COMP_RETURN_CAPS_COUNT];
+
+// Capability used to point to pair of capabilities when transitioning out of a
+// compartment via an intercept
+extern void* __capability sealed_redirect_cap;
+
+/* Data required to perform the transition for an intercepted function
+ */
+struct func_intercept {
+    char* func_name;
+    uintptr_t redirect_func;
+    void* __capability intercept_pcc;
+};
+
+/* This function expects the argument be passed in `x10`, rather than `x0`, as
+ * well as using `c29` as an argument for the DDC to transition to in order to
+ * allow the intercept to work. It is expected to be called only in very
+ * specific circumstances, and the signature is more illustrative than
+ * functional. As such, it shouldn't be called from a C context, as that will
+ * most likely break things.
+ */
+void intercept_wrapper(void* to_call_fn);
```

Comment:
Perhaps it's better to not give this a full prototype? i.e. remove the parameter?

Reply:
Incorrect git hash? At least the link isn't rendering for me.